src/components/sidebar/sidebar.component.tsx

```diff
@@ -1,134 +1,91 @@
+import { InlineLoading } from '@carbon/react';
 
 interface SidebarProps {
-  allowUnspecifiedAll: boolean;
   defaultPage: string;
-  handleClose: () => void;
-  hideFormCollapseToggle: () => void;
   isFormSubmitting: boolean;
-  mode: string;
+  sessionMode: SessionMode;
   onCancel: () => void;
-  pagesWithErrors: string[];
-  scrollablePages: Set<FormPage>;
-  selectedPage: string;
-  setValues: (values: unknown) => void;
-  values: object;
+  handleClose: () => void;
+  hideFormCollapseToggle: () => void;
 }
 
 const Sidebar: React.FC<SidebarProps> = ({
-  allowUnspecifiedAll,
   defaultPage,
-  handleClose,
-  hideFormCollapseToggle,
   isFormSubmitting,
-  mode,
+  sessionMode,
   onCancel,
-  pagesWithErrors,
-  scrollablePages,
-  selectedPage,
-  setValues,
-  values,
+  handleClose,
+  hideFormCollapseToggle,
 }) => {
   const { t } = useTranslation();
-  const pages: Array<FormPage> = Array.from(scrollablePages);
-
-  useEffect(() => {
-    if (defaultPage && pages.some(({ label, isHidden }) => label === defaultPage && !isHidden)) {
-      scrollIntoView(joinWord(defaultPage));
-    }
-  }, [defaultPage, scrollablePages]);
-
-  const unspecifiedFields = useMemo(
-    () =>
-      Object.keys(values).filter(
-        (key) => key.endsWith('-unspecified') && isEmpty(values[key.split('-unspecified')[0]]),
-      ),
-    [values],
-  );
-
-  const handleClick = (selected) => {
-    const activeId = joinWord(selected);
-    scrollIntoView(activeId);
-  };
-
-  const markAllAsUnspecified = useCallback(
-    (toggled) => {
-      const updatedValues = { ...values };
-      unspecifiedFields.forEach((field) => {
-        updatedValues[field] = toggled;
-      });
-      setValues(updatedValues);
-    },
-    [unspecifiedFields, values, setValues],
-  );
+  const { pages, pagesWithErrors, activePages, evaluatedPagesVisibility } = usePageObserver();
+  const { currentActivePage, requestPage } = useCurrentActivePage({
+    pages,
+    defaultPage,
+    activePages,
+    evaluatedPagesVisibility,
+  });
 
   return (
     <div className={styles.sidebar}>
-      {pages.map((page, index) => {
-        if (page.isHidden) return null;
+      {pages
+        .filter((page) => isPageContentVisible(page))
+        .map((page) => {
+          const isActive = page.id === currentActivePage;
+          const hasError = pagesWithErrors.includes(page.id);
+          return (
+            <div
+              className={classNames(styles.tab, {
+                [styles.activeTab]: isActive && !hasError,
+                [styles.errorTab]: hasError && !isActive,
+                [styles.activeErrorTab]: hasError && isActive,
+              })}>
+              <button
+                onClick={(e) => {
+                  e.preventDefault();
+                  requestPage(page.id);
+                }}>
+                <span>{page.label}</span>
+              </button>
+            </div>
+          );
+        })}
```

Comment:
This feels like the type of thing that might be better handled in a `useMemo()` hook. At the _least_, the top-level divs here need a `key` property.

Reply:
Better yet, we would refactor the div / button combo into its own component (it can be an internal component local to this file).

Reply:
>This feels like the type of thing that might be better handled in a useMemo() hook.

A more ideal approach would be memoizing this once in the form renderer and we expect the page observer to report only visible pages, but there is an issue:

Achieving this is likely to beg for the need of a state update of the entire form object every time a field changes, triggering the form to re-render every time the user interacts with a field. Currently, when a field's state changes, only the field and the flattened fields array state are updated (this doesn't trigger a re-render of the entire form).

But why would this be necessary? This boils down to how we determine the visibility of a page. A page is considered to have visible content if:

- The page itself is not hidden.
- At least one section within the page is visible.
- At least one question within each section is visible.

I think for now, we can deploy keys to the heat-map level elements, leveraging React's reconciliation algorithm and address the bigger issue in a different PR.

Reply:
Well, yeah, the whole need to re-render the form was kind of why I was suggesting local memoization, but I think I see your point: any field update can (potentially) trigger new fields to show and effect page visibility, and there isn't an obvious way to tie memoization to changing page visibility, so let's leave that aside.